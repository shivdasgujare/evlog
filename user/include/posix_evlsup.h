/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2001
 *
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */
#include <sys/un.h>
#ifndef _POSIX_EVLSUP_H_
#define _POSIX_EVLSUP_H_ 1

#ifdef __cplusplus
extern "C" {
#endif	/* __cplusplus */

typedef size_t HANDLE;

/* Name-value pair */
struct _evlNvPair {
	int	nv_value;
	char	*nv_name;
};

/* Values for faAccessFlags */
#define EVL_FACC_PRIVATE 0x1	/* facility's event recs are read-protected */
#define EVL_FACC_KERN 	 0x2

/* An entry from the facility registry */
typedef struct __evlFacility {
	posix_log_facility_t	faCode;
	char					*faName;
	int						faAccessFlags;
	char 					*faFilter;
	posix_log_query_t		*faQuery;
} _evlFacility;

/* Internal view of the info in /etc/evlog.d/facility_registry */
typedef struct __evlFacilityRegistry {
	int			frNFacilities;
	_evlFacility		*frHash;	/* hash table */
	int			frHashSize;	/* # slots in hash table */
} _evlFacilityRegistry;

extern void _evlGetNameByValue(const struct _evlNvPair table[], int value,
	char *name, size_t size, const char *dflt);
extern int _evlGetValueByName(const struct _evlNvPair table[], const char *name,
	int dflt);
extern int _evlGetValueByCIName(const struct _evlNvPair table[],
	const char *name, int dflt);
extern int _evlCIStrcmp(const char *s1, const char *s2);

extern struct _evlNvPair _evlLinuxFacilities[];
extern struct _evlNvPair _evlPosixFacilities[];
extern struct _evlNvPair _evlSeverities[];
extern struct _evlNvPair _evlAttributes[];
extern struct _evlNvPair _evlFormats[];
extern struct _evlNvPair _evlMgmtEventTypes[];

#define LOG_EVLOG_DIR EVLOG_STATE_DIR
#define LOG_EVLOG_CONF_DIR "/etc/evlog.d"
#define LOG_CURLOG_PATH LOG_EVLOG_DIR "/eventlog"
#define LOG_PRIVATE_PATH LOG_EVLOG_DIR "/privatelog"

#define EVL_INVALID_FACILITY ((posix_log_facility_t)-1)
#define EVL_INVALID_EVENT_TYPE (-1)

/*
 * The "age" pseudo-attribute -- must be distinct from POSIX_LOG_ENTRY_* in
 * posix_evlog.h.
 */
#define EVL_ENTRY_AGE 20
#define EVL_ENTRY_HOST 21

/* from posix2.c */
extern int _evlWrite(struct posix_log_entry *entry, const void *buf);
extern int _evlOpen(posix_logd_t *logdes, const char *path);
extern int _evlValidateLogdes(posix_logd_t logdes, int checkGeneration);
extern int _evlRead(posix_logd_t logdes, struct posix_log_entry *entry,
	void *buf, size_t bufLen);
extern int _evlClose(posix_logd_t fd);
extern int _evlRewind(posix_logd_t logdes);
extern int _evlSeekEnd(posix_logd_t logdes);
extern int _evlSeekFwd(posix_logd_t logdes, const posix_log_query_t *query);
extern int _evlSeekBkwd(posix_logd_t logdes, const posix_log_query_t *query);
extern int _evlFdWrite(int fd, const struct posix_log_entry *entry,
	const void *buf);
extern int _evlWriteLogHeader(int fd);
extern int _evlValidateQuery(posix_log_query_t *query, int nullOk);

/* facility registry functions */
extern void _evlLockFacReg(void);
extern void _evlUnlockFacReg(void);
extern int _evlBlockSignals(sigset_t *oldset);
extern void _evlRestoreSignals(sigset_t *oldset);
extern _evlFacilityRegistry *_evlReadFacilities(const char *facRegPath);
extern posix_log_facility_t _evlGetFacilityCodeByCIName(const char *name);
extern char *_evlGetFacilityName(posix_log_facility_t facNum, char *buf);
extern int _evlGetFacilityAccess(posix_log_facility_t facNum);
extern posix_log_query_t * _evlGetFacilityAccessQuery(posix_log_facility_t facNum, int * acc);
extern struct _evlNvPair *_evlSnapshotFacilities();
extern int _evlAddFacilityToRegistry(const char *fname, unsigned int flags,
	const char *filter);
extern unsigned int _evlCrc32(const unsigned char *data, int len);

/* posix1.c */
extern int _evlGetProcId();
extern int isAccess(int sd);
extern int _establishNonBlkConnection(const char * socketpath, struct sockaddr_un *sa, int nsec);
extern int _evlWriteEx(struct posix_log_entry *hdr, const char *buf);
extern int _evlEvaluateQuery(const posix_log_query_t *query,
	const struct posix_log_entry *entry, const void *buf);

#ifdef __cplusplus
}
#endif	/* __cplusplus */
#endif 	/* _POSIX_EVLSUP_H_ */
