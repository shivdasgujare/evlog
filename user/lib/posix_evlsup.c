/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2001
 *
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <sys/syslog.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <assert.h>
#include <stdlib.h>

#include "config.h"

#include "posix_evlog.h"
#include "posix_evlsup.h"
#include "evlog.h"

/*
 * These are the facilities defined by Linux's sys/syslog.h.
 * The facility registry must always contain at least these.
 */
struct _evlNvPair _evlLinuxFacilities[] = {
	{ LOG_AUTH,	"AUTH" },
	{ LOG_CRON,	"CRON" },
	{ LOG_DAEMON,	"DAEMON" },
	{ LOG_KERN,	"KERN" },
	{ LOG_LPR,	"LPR" },
	{ LOG_MAIL,	"MAIL" },
	{ LOG_NEWS,	"NEWS" },
	{ LOG_SYSLOG,	"SYSLOG" },
	{ LOG_USER,	"USER" },
	{ LOG_LOGMGMT,	"LOGMGMT" },

	{ LOG_UUCP,	"UUCP" },
	{ LOG_AUTHPRIV,	"AUTHPRIV" },
	{ LOG_FTP,	"FTP" },

	{ LOG_LOCAL0,	"LOCAL0" },
	{ LOG_LOCAL1,	"LOCAL1" },
	{ LOG_LOCAL2,	"LOCAL2" },
	{ LOG_LOCAL3,	"LOCAL3" },
	{ LOG_LOCAL4,	"LOCAL4" },
	{ LOG_LOCAL5,	"LOCAL5" },
	{ LOG_LOCAL6,	"LOCAL6" },
	{ LOG_LOCAL7,	"LOCAL7" },
	{ 0, 0 }
};

/*
 * These are the facilities defined by POSIX 1003.25.
 */
struct _evlNvPair _evlPosixFacilities[] = {
	{ LOG_AUTH,	"AUTH" },
	{ LOG_CRON,	"CRON" },
	{ LOG_DAEMON,	"DAEMON" },
	{ LOG_KERN,	"KERN" },
	{ LOG_LPR,	"LPR" },
	{ LOG_MAIL,	"MAIL" },
	{ LOG_NEWS,	"NEWS" },
	{ LOG_SYSLOG,	"SYSLOG" },
	{ LOG_USER,	"USER" },
	{ LOG_LOGMGMT,	"LOGMGMT" },
	{ LOG_UUCP,	"UUCP" },
	{ LOG_LOCAL0,	"LOCAL0" },
	{ LOG_LOCAL1,	"LOCAL1" },
	{ LOG_LOCAL2,	"LOCAL2" },
	{ LOG_LOCAL3,	"LOCAL3" },
	{ LOG_LOCAL4,	"LOCAL4" },
	{ LOG_LOCAL5,	"LOCAL5" },
	{ LOG_LOCAL6,	"LOCAL6" },
	{ LOG_LOCAL7,	"LOCAL7" },
	{ 0, 0 }
};

struct _evlNvPair _evlSeverities[] = {
	{ LOG_EMERG,	"EMERG" },
	{ LOG_ALERT,	"ALERT" },
	{ LOG_CRIT,	"CRIT" },
	{ LOG_ERR,	"ERR" },
	{ LOG_WARNING,	"WARNING" },
	{ LOG_NOTICE,	"NOTICE" },
	{ LOG_INFO,	"INFO" },
	{ LOG_DEBUG,	"DEBUG" },
	{ 0, 0 }
};

struct _evlNvPair _evlAttributes[] = {
	{ POSIX_LOG_ENTRY_RECID,	"recid" },
	{ POSIX_LOG_ENTRY_SIZE,	"size" },
	{ POSIX_LOG_ENTRY_FORMAT,	"format" },
	{ POSIX_LOG_ENTRY_EVENT_TYPE,	"event_type" },
	{ POSIX_LOG_ENTRY_FACILITY,	"facility" },
	{ POSIX_LOG_ENTRY_SEVERITY,	"severity" },
	{ POSIX_LOG_ENTRY_UID,	"uid" },
	{ POSIX_LOG_ENTRY_GID,	"gid" },
	{ POSIX_LOG_ENTRY_PID,	"pid" },
	{ POSIX_LOG_ENTRY_PGRP,	"pgrp" },
	{ POSIX_LOG_ENTRY_TIME,	"time" },
	{ POSIX_LOG_ENTRY_FLAGS,	"flags" },
	{ POSIX_LOG_ENTRY_THREAD,	"thread" },
	{ POSIX_LOG_ENTRY_PROCESSOR,	"processor" },
	{ 0, 0 }
};

struct _evlNvPair _evlFormats[] = {
	{ POSIX_LOG_NODATA,	"NODATA" },
	{ POSIX_LOG_BINARY,	"BINARY" },
	{ POSIX_LOG_STRING,	"STRING" },
	{ POSIX_LOG_PRINTF,	"PRINTF" },
	{ 0, 0 }
};

struct _evlNvPair _evlMgmtEventTypes[] = {
	{ POSIX_LOG_MGMT_TIMEMARK,		"MGMT_TIMEMARK" },
	{ POSIX_LOG_MGMT_STARTMAINT,	"MGMT_STARTMAINT" },
	{ POSIX_LOG_MGMT_ENDMAINT,		"MGMT_ENDMAINT" },
	{ POSIX_LOG_MGMT_STOPLOGGING,	"MGMT_STOPLOGGING" },
	{ POSIX_LOG_MGMT_STARTLOGGING,	"MGMT_STARTLOGGING" },
	{ 0, 0 }
};

void
_evlGetNameByValue(const struct _evlNvPair table[], int value, char *name, size_t size,
	const char *dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (nv->nv_value == value) {
			(void) strcpy(name, nv->nv_name);
			return;
		}
	}

	/* No match */
	if (dflt) {
		(void) strcpy(name, dflt);
	} else {
		snprintf(name, size, "%d", value);
	}
}

/* Look up value by name, case-sensitive */
int
_evlGetValueByName(const struct _evlNvPair table[], const char *name, int dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (!strcmp(nv->nv_name, name)) {
			return nv->nv_value;
		}
	}

	/* No match */
	return dflt;
}

/* Like strcmp, but case-insensitive */
int
_evlCIStrcmp(const char *s1, const char *s2)
{
	while (*s1 && toupper(*s1) == toupper(*s2)) {
		s1++;
		s2++;
	}
	return toupper(*s1) - toupper(*s2);
}

/* Look up value by name, case-insensitive */
int
_evlGetValueByCIName(const struct _evlNvPair table[], const char *name, int dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (!_evlCIStrcmp(nv->nv_name, name)) {
			return nv->nv_value;
		}
	}

	/* No match */
	return dflt;
}

/* Return the proc id */
int 
_evlGetProcId()
{
	int fd, ret;
	char buf[1024];
	char *last_token;
	size_t r;
	
	if ((fd = open("/proc/self/stat", O_RDONLY)) == -1) {
		fprintf(stderr, "Can't open stat\n");
		exit(1);
	}

	if ((r = read(fd, buf, 1024)) == -1) {
		close(fd);
		fprintf(stderr, "Read stat failed\n");
		exit(1);
	}
	buf[r]='\0';
	close(fd);
	/* The last token contains the processor id */

	last_token = (char *) strrchr(buf, ' ');
	assert(last_token != NULL);	
	return (int) strtol(last_token, (char **)NULL, 10);
}
