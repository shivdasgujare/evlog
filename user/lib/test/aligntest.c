/*
 * IBM Event Logging for Linux
 * Copyright (c) International Business Machines Corp., 2001
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or 
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include "posix_evlog.h"
#include "posix_evlsup.h"

/*
 * This spits out the same data as evl_log_write_test2.c, except that each
 * record is a single C struct.
 * All the templates are defined in align.t.
 */

main(int argc, char **argv)
{
	int status;
	posix_log_facility_t facility;
	int type;
	posix_log_severity_t severity;
	long larray[5]={30L,31L,32L,33L,34L};
	int iarray[5] = {1,2,3,4,5};

	long long llarray[5] = {10LL,11LL,12LL,13LL,14LL};
	unsigned long long ullarray[5]={20LL,21LL,22LL,23LL,24L};
	short sarray[5]={1,2,3,4,5};
	char carray[4]={'l','o','v','e'};
	unsigned char ucarray[4]={0xff,0x10,0xfe,0x20};
	long double ldarray[2]={2.33333, 3.333333};
	long double ld1 = 1.7676764;
	long double ld2 = 5.57575757;

	void *addr1 = (void*) 0xfeedf00d;
	void *addr2 = (void*) 0xfacef00d;
	void *addrarray[5] = { NULL, (void*)1, (void*)2, addr1, addr2 };
	
	facility = 8;
	severity = 3;
	
	type = 2001;
    {
	struct rec2001 {
		unsigned int	uint10;
		int	int10;
		int	iarray[5];
		short	short10;
		short	short11;
		/* string	love; */
		char	love[5];
		int	int20;
		int	iarrayAgain[5];
		long	longf0f0f0f0;
		char	charc;
	} rec;
	rec.uint10 = 10;
	rec.int10 = 10;
	memcpy(rec.iarray, iarray, 5*sizeof(int));
	rec.short10 = 10;
	rec.short11 = 11;
	strcpy(rec.love, "love");
	rec.int20 = 20;
	memcpy(rec.iarrayAgain, iarray, 5*sizeof(int));
	rec.longf0f0f0f0 = 0xf0f0f0f0;
	rec.charc = 'c';

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }

	type = 2002;
    {
	struct rec2002 {
		short		short9;
		long long	longlong56;
		unsigned long long	ulonglong56;
		int		int10;
		long long	llarray[3];
		/* string	peace; */
		char		peace[10];
		int		int20;
		int		iarray[5];
		long		longf0f0f0f0;
		char		charc;
	} rec;
	rec.short9 = 9;
	rec.longlong56 = 56LL;
	rec.ulonglong56 = 56LL;
	rec.int10 = 10;
	memcpy(rec.llarray, llarray, 3*sizeof(long long));
	strcpy(rec.peace, "peace");
	rec.int20 = 20;
	memcpy(rec.iarray, iarray, 5*sizeof(int));
	rec.longf0f0f0f0 = 0xf0f0f0f0;
	rec.charc = 'c';

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }

	type = 2003;
    {
    	struct rec2003 {
		long long		llarray[5];
		unsigned long long	ullarray[5];
		/* string linux */
		char			xlinux[10];
	} rec;
	memcpy(rec.llarray, llarray, 5*sizeof(long long));
	memcpy(rec.ullarray, ullarray, 5*sizeof(unsigned long long));
	strcpy(rec.xlinux, "linux");

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }

	type = 2004;
    {
    	struct rec2004 {
		int	iarray[5];
	} rec;
	memcpy(rec.iarray, iarray, 5*sizeof(int));

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }
	
	type = 2005;
    {
	struct rec2005 {
		short			short1;
		unsigned short		ushort2;
		long			long1111;
		unsigned long		ulong2222;
		unsigned long long	ulonglong10;
		long long		longlong10;
		float			float0_123;
		double			double9_999;
		long double		ld1;
	} rec;
	rec.short1 = 1;
	rec.ushort2 = 2;
	rec.long1111 = 1111L;
	rec.ulong2222 = 2222L;
	rec.ulonglong10 = 10LL;
	rec.longlong10 = 10LL;
	rec.float0_123 = 0.123;
	rec.double9_999 = 9.999;
	rec.ld1 = ld1;

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }
	
	type = 2006;
    {
	struct rec2006 {
		char		charA;
		unsigned char	ucharB;
		long double	ld2;
	} rec;
	rec.charA = 'A';
	rec.ucharB = 'B';
	rec.ld2 = ld2;

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }

	type = 2007;
    {
	struct rec2007 {
		long long		llarray[5];
		unsigned long long	ullarray[5];
		int			int10;
	} rec;
	memcpy(rec.llarray, llarray, 5*sizeof(long long));
	memcpy(rec.ullarray, ullarray, 5*sizeof(unsigned long long));
	rec.int10 = 10;

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }

	type = 2008;
    {
    	struct rec2008 {
		int		iarray[5];
		unsigned int	uiarray[5];
		short		sarray[5];
		unsigned short	usarray[5];
		char		carray[4];
		unsigned char	ucarray[4];
		unsigned char	peace[6];
		long		larray[5];
		unsigned char	xlinux[12];
	} rec;
	memcpy(rec.iarray, iarray, 5*sizeof(int));
	memcpy(rec.uiarray, iarray, 5*sizeof(int));
	memcpy(rec.sarray, sarray, 5*sizeof(short));
	memcpy(rec.usarray, sarray, 5*sizeof(short));
	memcpy(rec.carray, carray, 4*sizeof(char));
	memcpy(rec.ucarray, ucarray, 4*sizeof(unsigned char));
	strcpy(rec.peace, "peace");
	memcpy(rec.larray, larray, 5*sizeof(long));
	strcpy(rec.xlinux, "linux");

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }

	type = 2009;
    {
	struct rec2009 {
		int		int10;
		int		int11;
		int		int12;
		int		int13;
		int		int14;
		int		int15;
		long long	longlong10;
		long long	longlong11;
		long double	ldarray[2];
		long double	ld1;
		long double	ld2;
	} rec;
	rec.int10 = 10;
	rec.int11 = 11;
	rec.int12 = 12;
	rec.int13 = 13;
	rec.int14 = 14;
	rec.int15 = 15;
	rec.longlong10 = 10LL;
	rec.longlong11 = 11LL;
	memcpy(rec.ldarray, ldarray, 2*sizeof(long double));
	rec.ld1 = ld1;
	rec.ld2 = ld2;

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }

	type = 2010;
    {
    	struct rec2010 {
		void	*addr1;
		void	*addr2;
		void	*addrarray[5];
		void	*addr_feedface;
	} rec;
	rec.addr1 = addr1;
	rec.addr2 = addr2;
	memcpy(rec.addrarray, addrarray, 5*sizeof(void*));
	rec.addr_feedface = (void*) 0xfeedface;

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }

	type = 2020;
    {
    	struct foreignWord {
		char	word[13];
		int	language;
	};

	struct rec2020 {
		struct foreignWord	peace;
		struct foreignWord	love;
		struct foreignWord	slinux;
		struct foreignWord	elinux;
		char			equals[5];
		struct foreignWord	paz;
		struct foreignWord	amor;
	} rec;
	strcpy(rec.peace.word, "peace");
	strcpy(rec.love.word, "love");
	strcpy(rec.slinux.word, "Linux");
	strcpy(rec.elinux.word, "Linux");
	strcpy(rec.paz.word, "paz");
	strcpy(rec.amor.word, "amor");
	rec.peace.language = 'E';
	rec.love.language = 'E';
	rec.slinux.language = 'S';
	rec.elinux.language = 'E';
	rec.paz.language = 'S';
	rec.amor.language = 'S';
	strcpy(rec.equals, "is");

	evl_log_write(facility, type, severity, 0,
				"char[]", sizeof(rec), &rec,
				"endofdata");
    }
	exit(0);
}
