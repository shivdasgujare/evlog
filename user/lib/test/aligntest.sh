#!/bin/bash
#
TEST_TEMPLATE=aligntest.out

EVLTMPLPATH=$EVL_TESTDIR/templates
export EVLTMPLPATH

if [ "$USER" = root ]
then
	/sbin/evltc $EVLTMPLPATH/user/aligntest.t
else
echo >&2 "testevlwrite2.sh: Not running as root.  Assuming template source file"
echo >&2 "$EVLTMPLPATH/user/aligntest.t has already been compiled."
fi

logtime=`date +%s`
sleep 1
./aligntest
sleep 1
/sbin/evlview -f "(time>${logtime} && event_type >= 2001 && event_type <= 2100)" -S 'event type %event_type:d%:\n%data%' > aligntest.sh.out

diff aligntest.sh.out $TEST_TEMPLATE

if [ $? -eq "0" ]
then
	echo templates aligntest	:PASSED
else
	echo templates aligntest	:FAILED
fi
