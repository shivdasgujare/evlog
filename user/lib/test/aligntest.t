/*
 * Templates that correspond to the evl_log_write() calls in
 * evl_log_write_test2.c
 */
facility 8; event_type 2001;
aligned attributes {
	uint	uint10;
	int	int10;
	int	iarray[5]	"(%d )";
	short	short10;
	short	short11;
	char	love[5]		"%s";
	int	int20;
	int	iarrayAgain[5]	"(%d )";
	long	longf0f0f0f0	"%#lx";
	char	charc		"%c";
}
format
10=%uint10%
10=%int10%
1 2 3 4 5 =%iarray%
10=%short10%
11=%short11%
love=%love%
20=%int20%
1 2 3 4 5 =%iarrayAgain%
0xf0f0f0f0=%longf0f0f0f0%
c=%charc%
END


facility 8; event_type 2002;
aligned attributes {
	short		short9;
	longlong	longlong56;
	ulonglong	ulonglong56;
	int		int10;
	longlong	llarray[3]	"(%lld )";
	/* string	peace; */
	char		peace[10]	"%s";
	int		int20;
	int		iarray[5]	"(%d )";
	long		longf0f0f0f0	"%#lx";
	char		charc		"%c";
}
format
9=%short9%
56=%longlong56%
56=%ulonglong56%
10=%int10%
10 11 12 =%llarray%
peace=%peace%
20=%int20%
1 2 3 4 5 =%iarray%
0xf0f0f0f0=%longf0f0f0f0%
c=%charc%
END


facility 8; event_type 2003;
aligned attributes {
	longlong	llarray[5]	"(%lld )";
	ulonglong	ullarray[5]	"(%llu )";
	char		linux[10]	"%s";
}
format
10 11 12 13 14 =%llarray%
20 21 22 23 24 =%ullarray%
linux=%linux%
END


facility 8; event_type 2004;
aligned attributes {
	int	iarray[5]	"(%d )";
}
format
1 2 3 4 5 =%iarray%
END


facility 8; event_type 2005;
aligned attributes {
	short		short1;
	ushort		ushort2;
	long		long1111;
	ulong		ulong2222;
	ulonglong	ulonglong10;
	longlong	longlong10;
	float		float0_123	"%.3f";
	double		double9_999	"%.3f";
	ldouble		ld1	"%.7Lf";
}
format
1=%short1%
2=%ushort2%
1111=%long1111%
2222=%ulong2222%
10=%ulonglong10%
10=%longlong10%
0.123=%float0_123%
9.999=%double9_999%
1.7676764=%ld1%
END


facility 8; event_type 2006;
aligned attributes {
	char	charA	"%c";
	uchar	ucharB	"%c";
	ldouble	ld2	"%.8Lf";
}
format
A=%charA%
B=%ucharB%
5.57575757=%ld2%
END


facility 8; event_type 2007;
aligned attributes {
	longlong	llarray[5]	"(%lld )";
	ulonglong	ullarray[5]	"(%llu )";
	int		int10;
}
format
10 11 12 13 14 =%llarray%
20 21 22 23 24 =%ullarray%
10=%int10%
END


facility 8; event_type 2008;
aligned attributes {
	int	iarray[5]	"(%d )";
	uint	uiarray[5]	"(%u )";
	short	sarray[5]	"(%d )";
	ushort	usarray[5]	"(%u )";
	char	carray[4]	"(%c)";
	uchar	ucarray[4]	"(%#x )";
	uchar	peace[6]	"%s";
	long	larray[5]	"(%ld )";
	uchar	linux[12]	"%s";
}
format
1 2 3 4 5 =%iarray%
1 2 3 4 5 =%uiarray%
1 2 3 4 5 =%sarray%
1 2 3 4 5 =%usarray%
love=%carray%
0xff 0x10 0xfe 0x20 =%ucarray%
peace=%peace%
30 31 32 33 34 =%larray%
linux=%linux%
END


facility 8; event_type 2009;
aligned attributes {
	int		int10;
	int		int11;
	int		int12;
	int		int13;
	int		int14;
	int		int15;
	longlong	longlong10;
	longlong	longlong11;
	ldouble		ldarray[2]	"(%.6Lf )";
	ldouble		ld1		"%.7Lf";
	ldouble		ld2		"%.8Lf";
}
format
10=%int10%
11=%int11%
12=%int12%
13=%int13%
14=%int14%
15=%int15%
10=%longlong10%
11=%longlong11%
2.333330 3.333333 =%ldarray%
1.7676764=%ld1%
5.57575757=%ld2%
END


facility 8; event_type 2010;
aligned attributes {
	address	addr1;
	address	addr2;
	address	addrarray[5]	"(%p )";
	address	addr_feedface;
}
format
0xfeedf00d=%addr1%
0xfacef00d=%addr2%
(nil) 0x1 0x2 0xfeedf00d 0xfacef00d =%addrarray%
0xfeedface=%addr_feedface%
END


struct foreignWord;
aligned attributes {
	char	word[13]	"%s";
	int	language	"%c";
}
format
(%language%: %word%)
END


facility 8; event_type 2020;
aligned attributes {
	struct foreignWord	peace;
	struct foreignWord	love;
	struct foreignWord	slinux;
	struct foreignWord	elinux;
	char			equals[5]	"%s";
	struct foreignWord	paz;
	struct foreignWord	amor;
}
format
(E: peace) is (S: paz)=%peace% %equals% %paz%
(E: love) is (S: amor)=%love% %equals% %amor%
(E: Linux) is (S: Linux)=%elinux% %equals% %slinux%
