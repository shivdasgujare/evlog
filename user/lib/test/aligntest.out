event type 2001:
10=10
10=10
1 2 3 4 5 =1 2 3 4 5 
10=10
11=11
love=love
20=20
1 2 3 4 5 =1 2 3 4 5 
0xf0f0f0f0=0xf0f0f0f0
c=c
event type 2002:
9=9
56=56
56=56
10=10
10 11 12 =10 11 12 
peace=peace
20=20
1 2 3 4 5 =1 2 3 4 5 
0xf0f0f0f0=0xf0f0f0f0
c=c
event type 2003:
10 11 12 13 14 =10 11 12 13 14 
20 21 22 23 24 =20 21 22 23 24 
linux=linux
event type 2004:
1 2 3 4 5 =1 2 3 4 5 
event type 2005:
1=1
2=2
1111=1111
2222=2222
10=10
10=10
0.123=0.123
9.999=9.999
1.7676764=1.7676764
event type 2006:
A=A
B=B
5.57575757=5.57575757
event type 2007:
10 11 12 13 14 =10 11 12 13 14 
20 21 22 23 24 =20 21 22 23 24 
10=10
event type 2008:
1 2 3 4 5 =1 2 3 4 5 
1 2 3 4 5 =1 2 3 4 5 
1 2 3 4 5 =1 2 3 4 5 
1 2 3 4 5 =1 2 3 4 5 
love=love
0xff 0x10 0xfe 0x20 =0xff 0x10 0xfe 0x20 
peace=peace
30 31 32 33 34 =30 31 32 33 34 
linux=linux
event type 2009:
10=10
11=11
12=12
13=13
14=14
15=15
10=10
11=11
2.333330 3.333333 =2.333330 3.333333 
1.7676764=1.7676764
5.57575757=5.57575757
event type 2010:
0xfeedf00d=0xfeedf00d
0xfacef00d=0xfacef00d
(nil) 0x1 0x2 0xfeedf00d 0xfacef00d =(nil) 0x1 0x2 0xfeedf00d 0xfacef00d 
0xfeedface=0xfeedface
event type 2020:
(E: peace) is (S: paz)=(E: peace) is (S: paz)
(E: love) is (S: amor)=(E: love) is (S: amor)
(E: Linux) is (S: Linux)=(E: Linux) is (S: Linux)
